sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -262,27 +262,66 @@ case class GpuProjectAstExec(
 /**
  * Do projections in a tiered fashion, where earlier tiers contain sub-expressions that are
  * referenced in later tiers.  Each tier adds columns to the original batch corresponding
- * to the output of the sub-expressions.
+ * to the output of the sub-expressions.  It also removes columns that are no longer needed,
+ * based on inputAttrTiers for the current tier and the next tier.
  * Example of how this is processed:
  *   Original projection expressions:
  *   (((a + b) + c) * e), (((a + b) + d) * f), (a + e), (c + f)
  *   Input columns for tier 1: a, b, c, d, e, f  (original projection inputs)
  *   Tier 1: (a + b) as ref1
- *   Input columns for tier 2: a, b, c, d, e, f, ref1
+ *   Input columns for tier 2: a, c, d, e, f, ref1
  *   Tier 2: (ref1 + c) as ref2, (ref1 + d) as ref3
- *   Input columns for tier 3: a, b, c, d, e, f, ref1, ref2, ref3
+ *   Input columns for tier 3: a, c, e, f, ref2, ref3
  *   Tier 3: (ref2 * e), (ref3 * f), (a + e), (c + f)
  */
- case class GpuTieredProject(val exprSets: Seq[Seq[GpuExpression]]) extends Arm {
-
-  @tailrec
-  private def projectTier(boundExprs: Seq[Seq[GpuExpression]],
-      cb: ColumnarBatch, doClose: Boolean): ColumnarBatch = {
-    boundExprs match {
-      case Nil => {
-        cb
+ case class GpuTieredProject(exprTiers: Seq[Seq[GpuExpression]],
+    inputAttrTiers: Seq[AttributeSeq]) extends Arm {
+
+  // Determine which attributes in the first AttributeSeq can be skipped when
+  // building the list of attributes for the next AttributeSeq.
+  // Returns a list of boolean values the same size as the first
+  // sequence, where the values are true if the corresponding attribute can be skipped.
+  // The indices of these attributes correspond to the columns that were bound via
+  // GpuBindReferences.bindGpuReference.
+  private def getColumnSkips(inputTiers: Seq[AttributeSeq]): Array[Boolean] = inputTiers match {
+    case Nil => Array.emptyBooleanArray
+    case curTier :: Nil =>
+      // For the last tier, fill with all true (skip) values, because the output
+      // of the last tier does not include any input columns.
+      Array.fill(curTier.attrs.size)(true)
+    case curTier :: tail =>
+      val curAttrs = curTier.attrs
+      val nextAttrs = tail.head.attrs
+      // This is equivalent to:
+      // skipList = curAttrs.map(a => if (nextAttrs.contains(a)) false else true)
+      // but this should be faster
+      val skipList = new Array[Boolean](curAttrs.size)
+      var curIdx = 0
+      var nextIdx = 0
+      while (curIdx < curAttrs.size) {
+        if (nextAttrs(nextIdx) == curAttrs(curIdx)) {
```

Comment:
Absolutely.  But the order should be maintained.   This is why we have two indices, curIdx an nextIdx.  There will likely be additional attrs in curAttrs (outputs from this tier), but this loop is just looking at the which attrs (columns) in the current tier do not need to be carried forward to the next one.

Reply:
The question (which I bet is just because of how the tiers are defined) that could be an issue is where `nextAttrs` has less element than `curAttrs`. If that's not the case, then this is a non-issue. (i.e. if `nextAttrs` had less items then we could walk off the list in this loop?)

Reply:
> But it is important to note that all of the attrs/columns that are retained in the current tier will be at the front of the next tier. 

Thanks @jbrennan333, that's the key bit for me.